src/rest-server/test/v2/protocol.js

```diff
@@ -0,0 +1,470 @@
+name: caffe_example
+type: job
+version: 1.0.0
+contributor: OpenPAI
+prerequisites:
+  - protocolVersion: 2
+    name: caffe_example
+    type: dockerimage
+    version: 1.0.0
+    contributor : OpenPAI
+    description: caffe
+    uri: openpai/pai.example.caffe
+taskRoles:
+  valid_name:
+    instances: 1
+    completion:
+      minSucceededTaskCount: 1
+    dockerImage: caffe_example
+    resourcePerInstance:
+      cpu: 4
+      memoryMB: 8192
+    commands:
+      - ./examples/mnist/train_lenet.sh
+  `,
+
+  'should have required property \'commands\'': `
+protocolVersion: 2
+name: caffe_example
+type: job
+version: 1.0.0
+contributor: OpenPAI
+prerequisites:
+  - protocolVersion: 2
+    name: caffe_example
+    type: dockerimage
+    version: 1.0.0
+    contributor : OpenPAI
+    description: caffe
+    uri: openpai/pai.example.caffe
+taskRoles:
+  valid_name:
+    instances: 1
+    completion:
+      minSucceededTaskCount: 1
+    dockerImage: caffe_example
+    resourcePerInstance:
+      cpu: 4
+      memoryMB: 8192
+      gpu: 1
+  `,
+
+  'should be equal to one of the allowed values': `
+protocolVersion: 2
+name: caffe_mnist
+type: job
+version: 1.0.0
+contributor: OpenPAI
+prerequisites:
+  - protocolVersion: 2
+    name: caffe_example
+    type: dockerimage
+    version: 1.0.0
+    contributor : OpenPAI
+    description: caffe
+  `,
+};
+
+// unit tests for protocol in API v2
+describe('API v2 Unit Tests: protocol', () => {
+  // protocol schema test
+  it('test protocol schema for valid protocol', (done) => {
+    for (let validProtocolYAML of Object.values(validProtocolYAMLs)) {
+      const validProtocolJSON = yaml.safeLoad(validProtocolYAML);
+      expect(protocolSchema.validate(validProtocolJSON)).to.be.true;
+    }
+    done();
+  });
+  it('test protocol schema for invalid protocol', (done) => {
+    for (let invalidMessage of Object.keys(invalidProtocolYAMLs)) {
+      const invalidProtocolJSON = yaml.safeLoad(invalidProtocolYAMLs[invalidMessage]);
+      expect(protocolSchema.validate(invalidProtocolJSON)).to.be.false;
+      expect(protocolSchema.validate.errors[0].message).to.equal(invalidMessage);
+    }
+    done();
+  });
+  // protocol validation test
+  it('test protocol validation for valid protocol', async () => {
+    const req = {};
+    const res = {};
+    for (let validProtocolYAML of Object.values(validProtocolYAMLs)) {
+      req.body = {
+        protocol: validProtocolYAML,
+      };
+      await Promise
+        .resolve(protocolMiddleware.validate(req, res, () => {}))
+        .catch((err) => {
+          expect(err).to.be.null;
+        });
+    }
+  });
+  it('test protocol validation for invalid protocol', async () => {
+    const req = {};
+    const res = {};
+    for (let invalidMessage of Object.keys(invalidProtocolYAMLs)) {
+      req.body = {
+        protocol: invalidProtocolYAMLs[invalidMessage],
+      };
+      await Promise
+        .resolve(protocolMiddleware.validate(req, res, () => {}))
```

Comment:
`await chai.request(express().use(middleware)).get(...)` is better for unit testing middlewares, or, move validation function out of the middleware and test the function itself.

Reply:
updated